Adopt more smart pointers in dom/

cdumez opened the pull request:
#### 1cfcb963f8af7d7055a155939486947e5bfc2ce4
<pre>
Adopt more smart pointers in dom/
<a href="https://bugs.webkit.org/show_bug.cgi?id=263163">https://bugs.webkit.org/show_bug.cgi?id=263163</a>

Reviewed by Darin Adler.

* Source/WebCore/dom/AbortController.cpp:
(WebCore::AbortController::abort):
(WebCore::AbortController::protectedSignal const):
* Source/WebCore/dom/AbortController.h:
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::addSourceSignal):
(WebCore::AbortSignal::signalAbort):
(WebCore::AbortSignal::signalFollow):
* Source/WebCore/dom/ActiveDOMCallback.cpp:
(WebCore::ActiveDOMCallback::canInvokeCallback const):
(WebCore::ActiveDOMCallback::activeDOMObjectsAreSuspended const):
(WebCore::ActiveDOMCallback::activeDOMObjectAreStopped const):
* Source/WebCore/dom/ActiveDOMObject.cpp:
(WebCore::ActiveDOMObject::~ActiveDOMObject):
(WebCore::ActiveDOMObject::suspendIfNeeded):
(WebCore::ActiveDOMObject::queueTaskInEventLoop):
(WebCore::ActiveDOMObject::queueTaskToDispatchEventInternal):
(WebCore::ActiveDOMObject::queueCancellableTaskToDispatchEventInternal):
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::~Attr):
(WebCore::Attr::style):
(WebCore::Attr::detachFromElementWithValue):
* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::makeBoundaryPointBeforeNode):
(WebCore::makeBoundaryPointAfterNode):
(WebCore::treeOrder):
* Source/WebCore/dom/BroadcastChannel.cpp:
(WebCore::BroadcastChannel::MainThreadBridge::ensureOnMainThread):
(WebCore::BroadcastChannel::BroadcastChannel):
(WebCore::BroadcastChannel::protectedMainThreadBridge const):
(WebCore::BroadcastChannel::postMessage):
(WebCore::BroadcastChannel::close):
(WebCore::BroadcastChannel::isEligibleForMessaging const):
* Source/WebCore/dom/BroadcastChannel.h:
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::virtualCreate):
* Source/WebCore/dom/CharacterData.cpp:
(WebCore::CharacterData::~CharacterData):
(WebCore::canUseSetDataOptimization):
(WebCore::CharacterData::setData):
(WebCore::makeChildChange):
(WebCore::CharacterData::parserAppendData):
(WebCore::CharacterData::setDataAndUpdate):
(WebCore::CharacterData::notifyParentAfterChange):
(WebCore::CharacterData::dispatchModifiedEvent):
* Source/WebCore/dom/ChildListMutationScope.cpp:
(WebCore::ChildListMutationAccumulator::childAdded):
(WebCore::ChildListMutationAccumulator::willRemoveChild):
(WebCore::ChildListMutationAccumulator::enqueueMutationRecord):
(WebCore::ChildListMutationAccumulator::protectedTarget const):
* Source/WebCore/dom/ChildListMutationScope.h:
* Source/WebCore/dom/ChildNodeList.cpp:
(WebCore::ChildNodeList::~ChildNodeList):
* Source/WebCore/dom/ClassCollection.cpp:
(WebCore::ClassCollection::~ClassCollection):
* Source/WebCore/dom/ConstantPropertyMap.cpp:
(WebCore::ConstantPropertyMap::protectedDocument const):
(WebCore::ConstantPropertyMap::updateConstantsForSafeAreaInsets):
(WebCore::ConstantPropertyMap::didChangeSafeAreaInsets):
(WebCore::ConstantPropertyMap::updateConstantsForFullscreen):
(WebCore::ConstantPropertyMap::didChangeFullscreenInsets):
(WebCore::ConstantPropertyMap::setFullscreenAutoHideDuration):
* Source/WebCore/dom/ConstantPropertyMap.h:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertion):
(WebCore::ContainerNode::removeNodeWithScriptAssertion):
(WebCore::ContainerNode::removeDetachedChildren):
(WebCore::isChildTypeAllowed):
(WebCore::ContainerNode::insertBeforeCommon):
(WebCore::ContainerNode::replaceChild):
(WebCore::ContainerNode::replaceAll):
(WebCore::ContainerNode::rebuildSVGExtensionsElementsIfNecessary):
(WebCore::ContainerNode::removeChildren):
(WebCore::ContainerNode::appendChildWithoutPreInsertionValidityCheck):
(WebCore::ContainerNode::childrenChanged):
(WebCore::ContainerNode::cloneChildNodes):
(WebCore::dispatchChildInsertionEvents):
(WebCore::dispatchChildRemovalEvents):
(WebCore::ContainerNode::querySelectorAll):
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::notifyNodeInsertedIntoDocument):
(WebCore::notifyNodeInsertedIntoTree):
(WebCore::notifyChildNodeInserted):
(WebCore::notifyNodeRemovedFromDocument):
(WebCore::notifyNodeRemovedFromTree):
(WebCore::removeDetachedChildrenInContainer):

Canonical link: <a href="https://commits.webkit.org/269369@main">https://commits.webkit.org/269369@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2660fc300acfb5fa3d0b991eec780d87636db18

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22374 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22571 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24278 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20708 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26834 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22907 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21728 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22613 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22303 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/19429 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (cancelled)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25133 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20284 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20370 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20518 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24391 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/21046 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20499 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5325 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24735 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/21037 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/BroadcastChannel.cpp:
```diff
@@ -106,19 +106,21 @@ void BroadcastChannel::MainThreadBridge::ensureOnMainThread(Function<void(Page*)
     Ref protectedThis { *this };
     if (auto* document = dynamicDowncast<Document>(*context)) {
-        task(document->page());
+        CheckedPtr page = document->page();
```
Should we add `checkedPage` to make the idiom more elegant?

cdumez replied:
I assume to mean add a `checkedPage()` getter to Document? If so, why not. It seems to follow the pattern we use for `protectedX()`.

darinadler commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -592,14 +594,14 @@ ExceptionOr<void> ContainerNode::replaceChild(Node& newChild, Node& oldChild)
     ASSERT(refCount() || parentOrShadowHostNode());
 
-    Ref<ContainerNode> protectedThis(*this);
+    Ref protectedThis { *this };
```
This seems like the old idiom. In future we rely on the caller having protected `this`. But I guess it will always be too risky to remove this old protect code?

cdumez replied:
Correct, I am keeping those for now but this is the old idiom and it should not be needed anymore once our code has been fully ported to newer idioms. However, I think it is too early to remove those as our code is far from being fully ported.

cdumez committed:
Adopt more smart pointers in dom/
https://bugs.webkit.org/show_bug.cgi?id=263163

Reviewed by Darin Adler.

* Source/WebCore/dom/AbortController.cpp:
(WebCore::AbortController::abort):
(WebCore::AbortController::protectedSignal const):
* Source/WebCore/dom/AbortController.h:
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::addSourceSignal):
(WebCore::AbortSignal::signalAbort):
(WebCore::AbortSignal::signalFollow):
* Source/WebCore/dom/ActiveDOMCallback.cpp:
(WebCore::ActiveDOMCallback::canInvokeCallback const):
(WebCore::ActiveDOMCallback::activeDOMObjectsAreSuspended const):
(WebCore::ActiveDOMCallback::activeDOMObjectAreStopped const):
* Source/WebCore/dom/ActiveDOMObject.cpp:
(WebCore::ActiveDOMObject::~ActiveDOMObject):
(WebCore::ActiveDOMObject::suspendIfNeeded):
(WebCore::ActiveDOMObject::queueTaskInEventLoop):
(WebCore::ActiveDOMObject::queueTaskToDispatchEventInternal):
(WebCore::ActiveDOMObject::queueCancellableTaskToDispatchEventInternal):
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::~Attr):
(WebCore::Attr::style):
(WebCore::Attr::detachFromElementWithValue):
* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::makeBoundaryPointBeforeNode):
(WebCore::makeBoundaryPointAfterNode):
(WebCore::treeOrder):
* Source/WebCore/dom/BroadcastChannel.cpp:
(WebCore::BroadcastChannel::MainThreadBridge::ensureOnMainThread):
(WebCore::BroadcastChannel::BroadcastChannel):
(WebCore::BroadcastChannel::protectedMainThreadBridge const):
(WebCore::BroadcastChannel::postMessage):
(WebCore::BroadcastChannel::close):
(WebCore::BroadcastChannel::isEligibleForMessaging const):
* Source/WebCore/dom/BroadcastChannel.h:
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::virtualCreate):
* Source/WebCore/dom/CharacterData.cpp:
(WebCore::CharacterData::~CharacterData):
(WebCore::canUseSetDataOptimization):
(WebCore::CharacterData::setData):
(WebCore::makeChildChange):
(WebCore::CharacterData::parserAppendData):
(WebCore::CharacterData::setDataAndUpdate):
(WebCore::CharacterData::notifyParentAfterChange):
(WebCore::CharacterData::dispatchModifiedEvent):
* Source/WebCore/dom/ChildListMutationScope.cpp:
(WebCore::ChildListMutationAccumulator::childAdded):
(WebCore::ChildListMutationAccumulator::willRemoveChild):
(WebCore::ChildListMutationAccumulator::enqueueMutationRecord):
(WebCore::ChildListMutationAccumulator::protectedTarget const):
* Source/WebCore/dom/ChildListMutationScope.h:
* Source/WebCore/dom/ChildNodeList.cpp:
(WebCore::ChildNodeList::~ChildNodeList):
* Source/WebCore/dom/ClassCollection.cpp:
(WebCore::ClassCollection::~ClassCollection):
* Source/WebCore/dom/ConstantPropertyMap.cpp:
(WebCore::ConstantPropertyMap::protectedDocument const):
(WebCore::ConstantPropertyMap::updateConstantsForSafeAreaInsets):
(WebCore::ConstantPropertyMap::didChangeSafeAreaInsets):
(WebCore::ConstantPropertyMap::updateConstantsForFullscreen):
(WebCore::ConstantPropertyMap::didChangeFullscreenInsets):
(WebCore::ConstantPropertyMap::setFullscreenAutoHideDuration):
* Source/WebCore/dom/ConstantPropertyMap.h:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertion):
(WebCore::ContainerNode::removeNodeWithScriptAssertion):
(WebCore::ContainerNode::removeDetachedChildren):
(WebCore::isChildTypeAllowed):
(WebCore::ContainerNode::insertBeforeCommon):
(WebCore::ContainerNode::replaceChild):
(WebCore::ContainerNode::replaceAll):
(WebCore::ContainerNode::rebuildSVGExtensionsElementsIfNecessary):
(WebCore::ContainerNode::removeChildren):
(WebCore::ContainerNode::appendChildWithoutPreInsertionValidityCheck):
(WebCore::ContainerNode::childrenChanged):
(WebCore::ContainerNode::cloneChildNodes):
(WebCore::dispatchChildInsertionEvents):
(WebCore::dispatchChildRemovalEvents):
(WebCore::ContainerNode::querySelectorAll):
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::notifyNodeInsertedIntoDocument):
(WebCore::notifyNodeInsertedIntoTree):
(WebCore::notifyChildNodeInserted):
(WebCore::notifyNodeRemovedFromDocument):
(WebCore::notifyNodeRemovedFromTree):
(WebCore::removeDetachedChildrenInContainer):

Canonical link: https://commits.webkit.org/269369@main
